src/dotnet-svcutil/lib/src/CmdCredentialsProvider.cs

```diff
@@ -0,0 +1,306 @@
+#endif
+        }
+
+
+        private bool ValidateServerCertificate(object sender, X509Certificate cert, X509Chain chain, SslPolicyErrors sslPolicyErrors)
+        {
+            bool result = true;
+
+            HttpWebRequest request = sender as HttpWebRequest;
+            if (request != null && this.serviceUri != null && this.serviceUri.Authority == request.RequestUri.Authority)
+            {
+                result = sslPolicyErrors == SslPolicyErrors.None ? true : PromptUserOnInvalidCert(cert, sslPolicyErrors);
+            }
+
+            return result;
+        }
+
+        Dictionary<string, bool> validatedServerCerts = new Dictionary<string, bool>();
+
+        private bool PromptUserOnInvalidCert(X509Certificate cert, SslPolicyErrors sslPolicyErrors)
+        {
+            var certhash = cert.GetCertHashString();
+
+            if (!validatedServerCerts.Keys.Contains(certhash))
+            {
+                Console.WriteLine(string.Format(CultureInfo.InvariantCulture, SR.ErrServerCertFailedValidationFormat, sslPolicyErrors, FormatCertificate(cert)));
+                validatedServerCerts[certhash] = PromptEnterOrEscape(throwOnEscape: false);
+            }
+
+            return validatedServerCerts[certhash];
+        }
+
+        #endregion
+
+        #region ICloneable
+
+        public object Clone()
+        {
+            return new CmdCredentialsProvider();
+        }
+
+        #endregion
+
+        #region common functions
+
+        private static string FormatCertificate(X509Certificate cert, string certId = null)
+        {
+            var separator = "--------------------------------------------------------";
+            return separator + Environment.NewLine + certId + cert + separator;
+        }
+
+        private static bool PromptEnterOrEscape(bool throwOnEscape)
+        {
+            ConsoleKeyInfo keyInfo;
+
+            do
+            {
+                Console.WriteLine(SR.EnterOrEscapeMessage);
+                keyInfo = Console.ReadKey(intercept: true);
+            }
+            while (keyInfo.Key != ConsoleKey.Enter && keyInfo.Key != ConsoleKey.Escape);
+
+            Console.WriteLine();
+            if (keyInfo.Key == ConsoleKey.Escape && throwOnEscape)
+            {
+                throw new OperationCanceledException();
+            }
+
+            return keyInfo.Key == ConsoleKey.Enter;
+        }
+
+        public static string ReadUserInput(string prompt, bool isPassword = false)
+        {
+            ConsoleKeyInfo keyInfo;
+            var userInput = string.Empty;
+
+            Console.Write(prompt);
+
+            do
+            {
+                keyInfo = System.Console.ReadKey(intercept: true);
+                if (keyInfo.Key == ConsoleKey.Escape)
+                {
+                    Console.WriteLine();
+                    throw new OperationCanceledException();
+                }
+
+                if (!Char.IsControl(keyInfo.KeyChar))
+                {
```

Comment:
I'm curious, how does the backspace key behave? It looks like it wouldn't remove the character from the buffer. Also StringBuilder would be better to use here.

Reply:
You were right, the backspace key did not work as expected with this code. I changed it to stringbuilder and added some code to handle deleting the character from the buffer and from the screen when you press backspace.